# Exploring eBay Car Sales Data
This project uses a dataset of used cars from eBay _Kleinanzaigen_, a classifieds section of the German eBay website. The project's goal is to explore the dataset and find some patterns while practicing data cleansing skils.

The data was originally scraped and uploaded to Kaggle by user orgesleka. The original dataset is not available on Kaggle anymore and can be found for purchase [here](https://data.world/data-society/used-cars-data).

Data dictionary for the dataset:

- `dateCrawled` - When the ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing.
- `price` - The listed selling price of the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The type of vehicle.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The type of transmission.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `odometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelTyp`e - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreate`d - The date the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

## Dataset Exploration
The dataset needs to be loaded through Pandas to find out if there are any particular encodings or issues to fix to prepare the dataset for the later stages.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "latin1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

The shape of the dataset is 50,000 rows and 20 columns. Although, there is more information to be obtained using the `DataFrame.info()` and the `DataFrame.head()` functions.

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

An important observation is that the dataset needs to be cleaned. A few columns have null values. Some columns like `price`, `powerPS`, `monthOfRegistration`, `yearOfRegistration`, and `odometer` could benefit from taking a closer look into their data types for easier manipulation in the later stages of the data process. A mere example is the data type of the `price` column, which right now is an `object` type, meaning a string type. If we wanted to manipulate and analyze numbers, we could not proceed until converting the data type of similar columns. The `price` column's data type should be handled as a `float` type. Also, it is also worth taking a loot into entries denoting date and time, should they be converted into datetime objects for convenient data manipulation?

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dates are very specific, however, the names of the cars contain underscores `_`, some even descriptions of the capacitty of the engine in liters. The brand names appear to be written in lowercase. Depending on the use of the database, it may not be ideal to keep it that way but to change it to contain capital letters in the first letter of every single name. Another important observation is the name of the columns, they are not practical to read. Underscore notation normally is a better approach. Currently, camel case notation is being used.

## Cleaning Column Names
Let's convert the column names from camelcase to snakecase and change some of the column names based on the data dictionary in order to be more descriptive.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
column_names = autos.columns
columns_replacements = {
    "yearOfRegistration": "registration_year",
    "monthOfRegistration": "registration_month",
    "notRepairedDamage": "unrepaired_damage",
    "dateCreated": "ad_created",
    "offerType": "offer_type",
    "dateCrawled": "date_crawled",
    "vehicleType": "vehicle_type",
    "powerPS": "power_ps",
    "fuelType": "fuel_type",
    "nrOfPictures": "n_pictures",
    "postalCode": "postal_code",
    "lastSeen": "last_seen"
}
autos.rename(columns_replacements, inplace = True, axis = "columns")

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  n_pictures  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

All the columns received some changes in their names:
1. The columns were converted from camelcase to snakecase (use of underscores).
2. A few names received more descriptive names according to the information they represent, e.g. `yearOfRegistration` was rewritten as `registration_year`.

## Initial Exploration and Cleaning
The goal is to explore the data to recognize potential changes needed to be done. We will look for:
- Text columns where all or almost all the values are the same. Given they become useless for analysis, they may be dropped.
- Examples of numeric data stored as text which can be cleaned and converted.

Useful methods to explore the data:
- `DataFrame.describe()` (with `include='all'` to get both categorical and numeric columns)
- `Series.value_counts()` and `Series.head()` if any columns need a closer look.

In [8]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  n_pictures   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Values in columns like `seller` and `offer_type` are likely to be dropped due to the frequency of repeated values. The columns possess the same values in almost all the rows; In `seller`, the value `privat` is repeated 49,999 times out of the 50,000 entries. The case is similar for the `offer_type` column withm the value `Angebot`. The same may apply to the column `n_pictures`, where all equal to `0`.

In [59]:
autos.drop(columns = ["seller", "offer_type", "n_pictures"])

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0       5000.0  control          bus               2004    manuell       158   
1       8500.0  control    limousine               1997  automatik       286   
2       8990.0     test    limousine               2009    manuell       102   
3       4350.0  control   kleinwagen               2007  automatik        71   
4       1350.0     test        kombi               2003    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49995  24900.0  control    limousine               2011  automatik       239   
49996   1980.0  control       cabrio               1996    manuell        75   
49997  13200.0     test       cabrio               2014  automatik        69   
49998  22900.0  control        kombi               2013    manuell       150   
49999   1250.0  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

The columns `odometer`, `fuel_type`, `unrepaired_damage`, and `gearbox` contain a high frequency count, almost past half the 50,000 entry count. It may not be closed to the overall amount of 50,000, however, such a high frequency count should be explored in detail.

Some columns have numeric values store as `string` data types and could be cleaned to produce more useful insights. Such is the case of `date_crawled`, `price`, and `odometer`.

In [53]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  registration_year  \
0      privat      Offer   5000.0  control          bus               2004   
1      privat      Offer   8500.0  control    limousine               1997   
2      privat      Offer   8990.0     test    limousine               2009   
3      privat      Offer   4350.0  control   kleinwagen               2007   
4      privat      Offer   1350.0     test        kombi               2003   
...       ...        ...      ...      ...          ...                ...   
49995  privat      Offer  24900.0  control    limousine               2011   
49996  privat      Offer   1980.0  control       cabrio               1996   
49997  privat      Offer  13200.0     test       cabrio               2014   
49998  privat      Offer  22900.0  control        kombi               2013   
49999  privat      Offer   1250.0  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  n_pictures  \
0         peugeot              nein  2016-03-26 00:00:00           0   
1             bmw              nein  2016-04-04 00:00:00           0   
2      volkswagen              nein  2016-03-26 00:00:00           0   
3           smart              nein  2016-03-12 00:00:00           0   
4            ford              nein  2016-04-01 00:00:00           0   
...           ...               ...                  ...         ...   
49995        audi              nein  2016-03-27 00:00:00           0   
49996        opel              nein  2016-03-28 00:00:00           0   
49997        fiat              nein  2016-04-02 00:00:00           0   
49998        audi              nein  2016-03-08 00:00:00           0   
49999        opel              nein  2016-03-13 00:00:00           0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 14:38:50  
...           

Some of the words contained within the dataset are in german, let us start by fixing the words for classifying the `offer_type` column by translating it. Let us take a look into said column.The columns `odometer`, `fuel_type`, `unrepaired_damage`, and `gearbox` contain a high frequency count, almost past half the 50,000 entry count. It may not be closed to the overall amount of 50,000, however, such a high frequency count should be explored in detail.

Some columns have numeric values store as `string` data types and could be cleaned to produce more useful insights. Such is the case of `date_crawled`, `price`, and `odometer`.

### Further Anlysis
The `odometer` column  displays numeric values delimited by a `,`and a `km` string concatenated. Fortunately, there are no special cases in the column's values, therefore, the cleaning process is pretty typical; The values, once cleaned, will be converted into `integer` types.

In [11]:
autos["odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [12]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In the case of the `price` column has a string format where the `$` and `,` characters are found. These characters are to be removed and the data types will be converted to `float`.

In [13]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [14]:
autos["price"].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: price, Length: 2357, dtype: int64

### Further Cleaning
We will start the cleaning with the `price` column.

In [15]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "").astype(float)

/tmp/ipykernel_202/1509714250.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$", "")


Now, it is time to explore the results.

In [16]:
autos["price"].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

Proceeding with the `odometer` column, we will replace the non-numeric characters in the values of the column and rename the column from `odometer` to `odometer_km`, making it more descriptive.

In [17]:
autos["odometer"] = autos["odometer"].str.replace(",", "")
autos["odometer"] = autos["odometer"].str.replace("km", "").astype(int)

Let us proceed to explore the results.

In [18]:
autos.rename({"odometer": "odometer_km"}, inplace = True, axis = 1)

In [19]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   seller              50000 non-null  object 
 3   offer_type          50000 non-null  object 
 4   price               50000 non-null  float64
 5   abtest              50000 non-null  object 
 6   vehicle_type        44905 non-null  object 
 7   registration_year   50000 non-null  int64  
 8   gearbox             47320 non-null  object 
 9   power_ps            50000 non-null  int64  
 10  model               47242 non-null  object 
 11  odometer_km         50000 non-null  int64  
 12  registration_month  50000 non-null  int64  
 13  fuel_type           45518 non-null  object 
 14  brand               50000 non-null  object 
 15  unrepaired_damage   40171 non-null  object 
 16  ad_c

## Exploring the Odomoter and Price Columns
Let us continue exploring the data, specifically looking for data that does not look right. We will start analyzing the `odometer_km` and `price` columns.

We will analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove as much as some other tools like looking at the shape of the data frame, order it, etc.

Let us start with the `odometer_km` column by looking at the unique values.

In [20]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

Now, let us take a look at the count of unique values.

In [21]:
autos["odometer_km"].unique().shape

(13,)

It is time to view the statistical description of the column: Min, Max, Mean, Median, etc.

In [22]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Let's get the frequency values of each value withing the column.

In [23]:
autos["odometer_km"].value_counts().sort_index(ascending = False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

Now, we should proceed to explore the `price` column.

In [24]:
autos["price"].shape

(50000,)

There is a high probability to obtain some interesting information in the following section.

In [25]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

The description of the `price` column shows that some prices are equal to `0`. It means that some of the cars may be depicted as "free".

In [26]:
autos["price"].value_counts().sort_index(ascending = False)

99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price, Length: 2357, dtype: int64

In [27]:
sorted(autos["price"], reverse = True)[:10]

[99999999.0,
 27322222.0,
 12345678.0,
 12345678.0,
 12345678.0,
 11111111.0,
 11111111.0,
 10000000.0,
 3890000.0,
 1300000.0]

Up to this point, the `odometer_km` column contains no perceivable outliers. Although, in the case of the `price` column some values definitely raise some questions:
- How is it possible to have a car worth 99 million Euros? Some of the most expensive cars barely cost 2 or 3 million Euros.
- How can a car cost 0 Euros in website where offers are for purchase?
- Would a car cost 5 Euros?
- Do the cars offered at abnormal prices normally cost that much? 
- What are the names of the mentioned cars?

In countries like Mexico, governmental organisms like the PROFECO, an organisation in charge of looking for the rights of consumers, make sure that prices are respected when the buyer has purchased or looked at a product being offered with a certain price tag. Nevertheless, the regulations may not apply in the same manner for every country. Let us remember that this website is the german version of eBay. 

In this project, the criteria will stick to the common prices at which a certain car may be offered at. An example is to look for the name of a car and look for its price in the used cars market. Let us also suppose that no Premium or Luxury car can be found normally at less than 100,000 euros or that no car with a fuctional engine can cost 5 Euros.

To proceed into the removal of outliuers, we must take a look into the most expensive cars and see which cars and their prices would not match each other normally. The first step is to look into cars worth more than 100,000 euros.

In [28]:
autos.loc[autos["price"] > 100000, ["name", "price"]].sort_values("price", ascending = False)

name       price
39705                        Tausch_gegen_gleichwertiges  99999999.0
42221                                  Leasinguebernahme  27322222.0
39377                        Tausche_volvo_v40_gegen_van  12345678.0
27371                                         Fiat_Punto  12345678.0
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  12345678.0
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  11111111.0
24384                            Schlachte_Golf_3_gt_tdi  11111111.0
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  10000000.0
47634                                        Ferrari_FXX   3890000.0
7814                                         Ferrari_F40   1300000.0
22947                         Bmw_530d_zum_ausschlachten   1234566.0
43049                                      2_VW_Busse_T3    999999.0
514                     Ford_Focus_Turnier_1.6_16V_Style    999999.0
37585                                Volkswagen_Jetta_GT    999990.0
36818                                        Porsche_991    350000.0
14715                 Rolls_Royce_Phantom_Drophead_Coupe    345000.0
34723  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...    299000.0
35923  Porsche_911_Targa_Exclusive_Edition__1_von_15_...    295000.0
12682  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...    265000.0
47337                                    BMW_Z8_roadster    259000.0
38299                                Glas_BMW_mit_Wasser    250000.0
37840                                        Porsche_997    220000.0
40918                             Porsche_911_991_GT3_RS    198000.0
43668  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...    197000.0
28090  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...    194000.0
20351                                    BMW_Z8_roadster    190000.0
17140  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT    180000.0
11433  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...    175000.0
32840  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...    169999.0
18509                                    Porsche_991_GT3    169000.0
22673  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...    163991.0
45387  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...    163500.0
10500                                        Porsche_991    155000.0
33638                                 Porsche_911_GT3_RS    151990.0
49668         Porsche_911__993__Carrera_S_"SONDERMODELL"    145000.0
32185  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...    139997.0
2454                                     Porsche_911_GT3    137999.0
14268   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011    135000.0
49815  SUCHE_TIPPS___Ford_Mustang_Shelby_GT_350_500_K...    130000.0
1878                                   Porsche_911_Turbo    129000.0
8232                Porsche_993_S_Schalter_BRD_neuwertig    128000.0
44406  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...    120000.0
2751                                Porsche_911___993_4S    120000.0
43282               Porsche_911_Carrera_4S_Cabriolet_PDK    119900.0
38814  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...    119500.0
33884               Porsche_911_Carrera_4S_Cabriolet_PDK    116000.0
21783               Porsche_911_Carrera_4S_Cabriolet_PDK    115991.0
7402   Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...    115000.0
22060  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...    114400.0
49391  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...    109999.0
17540   Porsche_930_Turbo_classic_analytics_Gutachten_2+    105000.0
16964                    Bentley_Continental_Supersports    105000.0
29286  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...    104900.0

Some rows contain unrealistic prices and names that do not include car models' names. An example is "Tausch_gegen_gleichwertiges", which means "Exchange per equivalent", meaning it is not a car. Adding up to this case example, the price is unrealistic: 99 million euros. No commercial luxury car costs that much in the world. 

Concluding from this analysis, on the higher end, removing cars costing above 4 million euros would be the best course of action, as those entries contain unrealistic scenarios (outliers).

To follow up appropriately, we must revise an article written by the renowned law firm Wiler Hale called "[Are sellers bound by mistakes in online advertisements?](https://www.wilmerhale.com/insights/publications/are-sellers-bound-by-mistakes-in-online-advertisements-june-30-2003)". The article explains that in Germany, depending on the case scenario, the court may or may not make a customer's offer valid. This means that the court takes into account on each case the information avaiable to determine wether a purchase's price should be respected or not. If a case like offering a car at the price of 0 euros comes up, the court may take it like a common sense scenario where the validity of the claim might not proceed. In their view, no car costs 0 eruos in a used cars website. Taking this into account, the criterion selected to leave out any cars will be include any car costing equals or above 1 euro.

Let us now revise the bottom list in terms of prices and their entries. What odd scenarios can be found? 

In [29]:
autos["price"].value_counts().sort_index(ascending = False).tail(200)

760.0       8
755.0       1
750.0     433
749.0      23
745.0       2
         ... 
5.0         2
3.0         1
2.0         3
1.0       156
0.0      1421
Name: price, Length: 200, dtype: int64

What kind of cars cost 1 euro?

In [30]:
autos[autos["price"] == 1]

date_crawled                                               name  \
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
344    2016-03-22 11:25:18  Verkaufe_hier_mein_mein_schoener_honda_crx_in_...   
791    2016-04-05 09:52:11                              Opel_Vectra_in_teilen   
981    2016-03-10 00:56:12                            Verkaufen_oder_tauschen   
1115   2016-03-24 22:36:58                           Suche_Passat_VR6_Variant   
...                    ...                                                ...   
49240  2016-03-10 09:54:48   Suche_7_Sitzer_Espace_oder?_Biete_s.Beschreibung   
49377  2016-03-20 00:58:36             Tausche_BMW_318i_Facelift_Model_143_ps   
49528  2016-03-28 19:57:41       BMW_E36__316i_Compact_M_Paket_"Schlachtfest"   
49728  2016-03-25 16:00:11                     Ford_Focus_1.4_16V_Bastlerauto   
49845  2016-03-21 00:50:15  Schlachte_VW_Sharan_vr6_Automatik___no_GTI_16V...   

       seller offer_type  price   abtest vehicle_type  registration_year  \
55     privat      Offer    1.0     test          NaN               2017   
344    privat      Offer    1.0     test          NaN               2000   
791    privat      Offer    1.0  control        kombi               1999   
981    privat      Offer    1.0  control        kombi               2003   
1115   privat      Offer    1.0  control          NaN               1995   
...       ...        ...    ...      ...          ...                ...   
49240  privat      Offer    1.0     test        kombi               1999   
49377  privat      Offer    1.0  control    limousine               2002   
49528  privat      Offer    1.0  control          NaN               2016   
49728  privat      Offer    1.0     test          NaN               2016   
49845  privat      Offer    1.0     test          NaN               2000   

         gearbox  power_ps     model  odometer_km  registration_month  \
55     automatik       224  e_klasse       125000                   7   
344      manuell       125    andere         5000                   0   
791      manuell       101    vectra       150000                   1   
981      manuell       150       3er       150000                   9   
1115     manuell         0    passat       150000                   0   
...          ...       ...       ...          ...                 ...   
49240    manuell         0       NaN       150000                   4   
49377    manuell       143       3er       150000                   8   
49528    manuell       102       3er       150000                  12   
49728    manuell         0     focus       150000                   0   
49845  automatik       174    sharan         5000                   6   

      fuel_type          brand unrepaired_damage           ad_created  \
55       benzin  mercedes_benz              nein  2016-03-06 00:00:00   
344         NaN          honda               NaN  2016-03-22 00:00:00   
791      benzin           opel               NaN  2016-04-05 00:00:00   
981      diesel            bmw              nein  2016-03-10 00:00:00   
1115     benzin     volkswagen               NaN  2016-03-24 00:00:00   
...         ...            ...               ...                  ...   
49240    benzin            bmw               NaN  2016-03-10 00:00:00   
49377    benzin            bmw               NaN  2016-03-19 00:00:00   
49528    benzin            bmw               NaN  2016-03-28 00:00:00   
49728    benzin           ford                ja  2016-03-25 00:00:00   
49845    benzin     volkswagen              nein  2016-03-20 00:00:00   

       n_pictures  postal_code            last_seen  
55              0        22111  2016-03-08 05:45:44  
344             0        65391  2016-04-06 18:15:57  
791             0        59394  2016-04-07 13:16:37  
981             0        58840  2016-04-06 01:16:59  
1115            0        38556  2016-04-06 03:45:42  
...           ...          ...                  ..

Now, proceeding with the cleaning of the data.

In [31]:
autos_cleaned = autos[autos["price"].between(1, 4000000)]

In [32]:
autos_cleaned

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  registration_year  \
0      privat      Offer   5000.0  control          bus               2004   
1      privat      Offer   8500.0  control    limousine               1997   
2      privat      Offer   8990.0     test    limousine               2009   
3      privat      Offer   4350.0  control   kleinwagen               2007   
4      privat      Offer   1350.0     test        kombi               2003   
...       ...        ...      ...      ...          ...                ...   
49995  privat      Offer  24900.0  control    limousine               2011   
49996  privat      Offer   1980.0  control       cabrio               1996   
49997  privat      Offer  13200.0     test       cabrio               2014   
49998  privat      Offer  22900.0  control        kombi               2013   
49999  privat      Offer   1250.0  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  n_pictures  \
0         peugeot              nein  2016-03-26 00:00:00           0   
1             bmw              nein  2016-04-04 00:00:00           0   
2      volkswagen              nein  2016-03-26 00:00:00           0   
3           smart              nein  2016-03-12 00:00:00           0   
4            ford              nein  2016-04-01 00:00:00           0   
...           ...               ...                  ...         ...   
49995        audi              nein  2016-03-27 00:00:00           0   
49996        opel              nein  2016-03-28 00:00:00           0   
49997        fiat              nein  2016-04-02 00:00:00           0   
49998        audi              nein  2016-03-08 00:00:00           0   
49999        opel              nein  2016-03-13 00:00:00           0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 14:38:50  
...           

In [33]:
autos_cleaned["price"].describe()

count    4.857100e+04
mean     6.082245e+03
std      2.278698e+04
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.890000e+06
Name: price, dtype: float64

The outliers were removed based on the research performed and the set criteria. The minimum value found in the `price` column is equal to 1 euro, as for the maximum value is 3.89 million euros. The remaining values are left in accordance with the criteria looking for the healthiest possible dataset. 

Even though the remaining data has been cleaned in terms of numeric values, there are many data points that are not even cars or simply refer to random sentences appointed are cars contained within the `name` column.

This is an example of row which should be a car, but it's a random sentence written in german:

In [34]:
autos_cleaned.loc[101, "name"]

'Schnaepchen_in_einem_Jahr_OLDTEIMER_KENNZEICHEN'

## Exploring the Date Columns


Moving on to the date columns, we will try to understand the date range the data covers.

Referenciating the data dictionary we may recognize which of the columns have been created by the website itself or by the crawler:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The columns `date_crawled`, `last_seen`, and `ad_created` are identified by pandas as strings given the context they were obtained on. These columns are needed to be converted into numerical values to handle them appropriately.

Let us first take a look into the form these columns' values are currently being formatted.

In [35]:
autos[["date_crawled", "last_seen", "ad_created"]].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

Now, we must analyze the range of the date by generating a distribution of the dates as percentages. This is the distribution for the `date_crawled` column.

In [36]:
autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

The dates range from `2016-03-05` to `2016-04-07`. This looks appropriate when putting into perspective the years of creation of the posts against the dates the data was crawled on.

In the case of the `ad_created` column, this is the distribution.

In [37]:
autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

The dates range from `2015-06-11` to `2016-04-07`. No outliers are perceived in the column.

Last but not least, this is the distribution of the `last_seen` column.

In [38]:
autos["last_seen"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

The dates range from `2016-03-05` to `2016-04-07`.

The numeric columns `registration_month` and `registration_year` also need exploration.

In [39]:
autos[["registration_month", "registration_year"]].describe()

registration_month  registration_year
count        50000.000000       50000.000000
mean             5.723360        2005.073280
std              3.711984         105.712813
min              0.000000        1000.000000
25%              3.000000        1999.000000
50%              6.000000        2003.000000
75%              9.000000        2008.000000
max             12.000000        9999.000000

From this line of code we have found that there are some outliers in the information:
- `registration_year` - Theres is a value equal to year `1000`, before cars were invented and a year `9999`, a year which has not happened yet.

## Dealing with Incorrect Registration Year Data


One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:
- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [40]:
autos_cleaned.loc[autos_cleaned["registration_year"] > 1000, ["registration_year", "registration_month"]].sort_values("registration_year")

registration_year  registration_month
49283               1001                   0
24511               1111                   0
10556               1800                   2
32585               1800                   2
3679                1910                   0
...                  ...                 ...
25003               8888                   0
49910               9000                   3
8012                9999                   0
38076               9999                   0
33950               9999                   0

[48570 rows x 2 columns]

The first pattented car was created in **1886**, meaning that we could use said year as the lower limit for our selection. As for the upper limit, the year that the dataset was obtained was **2016**, implying the impossibility of adding a car with a newer registry.

In [41]:
autos_cleaned = autos_cleaned[autos_cleaned["registration_year"].between(1886, 2016)]

In [42]:
autos_cleaned["registration_year"].sort_values()

28693    1910
22659    1910
45157    1910
30781    1910
3679     1910
         ... 
5140     2016
48828    2016
7809     2016
16546    2016
17629    2016
Name: registration_year, Length: 46687, dtype: int64

## Exploring Price by Brand
Let's explore the `brand` column and select carefully which ones to include by aggregating brands per mean price.

Let us first get a list with all the brands in the dataset.

In [43]:
brands = autos_cleaned["brand"].unique()
brands

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [44]:
len(brands)

40

This list will be used to loop over the brands in the dataset and calculate their average prices.

In [45]:
brands_avg_price = {}
for brand in brands:
    brands_avg_price[brand] = autos_cleaned.loc[autos_cleaned["brand"] == brand, "price"].mean()

This are all the resultant brands and their average prices.

In [46]:
brands_avg_price = dict(sorted(brands_avg_price.items(), key = lambda item: item[1], reverse = True))
brands_avg_price

{'porsche': 45643.93706293706,
 'sonstige_autos': 23567.513043478262,
 'land_rover': 19108.091836734693,
 'jeep': 11650.5,
 'jaguar': 11635.493150684932,
 'mini': 10613.459657701711,
 'audi': 9336.687453600594,
 'mercedes_benz': 8628.450366422385,
 'bmw': 8571.480147917478,
 'chevrolet': 6684.139097744361,
 'skoda': 6368.0,
 'kia': 5982.330303030303,
 'dacia': 5915.528455284553,
 'volkswagen': 5604.071269261963,
 'hyundai': 5365.254273504273,
 'toyota': 5167.091062394604,
 'volvo': 4946.501170960188,
 'nissan': 4743.40252454418,
 'seat': 4397.230949589683,
 'mazda': 4112.596614950635,
 'honda': 4107.857923497268,
 'suzuki': 4096.732851985559,
 'alfa_romeo': 4087.690322580645,
 'ford': 4054.6930147058824,
 'subaru': 3953.29,
 'citroen': 3779.1391437308866,
 'smart': 3580.2239031770046,
 'chrysler': 3465.743902439024,
 'mitsubishi': 3394.5729166666665,
 'lancia': 3376.22,
 'saab': 3211.6493506493507,
 'peugeot': 3094.0172290021537,
 'opel': 2975.2419354838707,
 'fiat': 2813.748538011696,

Aggregating this type of data is only useful if it represents the complete population appropriately. Given that this population's prices do not vary drastically, we may select the top 20 brands. An interesting point to notice is that the lower end prices do not vary and tend to be quite small in perspective to the rest of the column.

Let us proceed to filter the top 20 brands.

In [47]:
brands_price = {}
count = 0
for brand, price in brands_avg_price.items():
    if count < 20:
        brands_price[brand] = price
    count += 1

In [48]:
brands_price

{'porsche': 45643.93706293706,
 'sonstige_autos': 23567.513043478262,
 'land_rover': 19108.091836734693,
 'jeep': 11650.5,
 'jaguar': 11635.493150684932,
 'mini': 10613.459657701711,
 'audi': 9336.687453600594,
 'mercedes_benz': 8628.450366422385,
 'bmw': 8571.480147917478,
 'chevrolet': 6684.139097744361,
 'skoda': 6368.0,
 'kia': 5982.330303030303,
 'dacia': 5915.528455284553,
 'volkswagen': 5604.071269261963,
 'hyundai': 5365.254273504273,
 'toyota': 5167.091062394604,
 'volvo': 4946.501170960188,
 'nissan': 4743.40252454418,
 'seat': 4397.230949589683,
 'mazda': 4112.596614950635}

Most of the values of the prices are less than 10,000 euros. There is an extreme polarization of the values only in three brands: Porsche, Land Rover, and Jeep. The word _Sonstige_ in german is "others" in English. Basically, it refers to a certain pack of cars not entering the classification of the most common brands.

## Storing Aggregate Data in a DataFrame
Looking at the past screen, we may investigate deeper into the relationship between the top brands, their prices and if there is another kind of relationship to another factor like mileage. For this, it is needed to create an average list to relate the top brands to the average mileage per brand.

In [49]:
brands_mileage = {}
for brand in brands_price:
    brands_mileage[brand] = autos_cleaned.loc[autos_cleaned["brand"] == brand, "odometer_km"].mean()

In [50]:
brands_mileage = dict(sorted(brands_mileage.items(), key = lambda item: item[1]))
brands_mileage

{'dacia': 84268.29268292683,
 'mini': 88105.13447432763,
 'sonstige_autos': 89684.78260869565,
 'porsche': 96853.14685314686,
 'chevrolet': 99191.72932330827,
 'hyundai': 106442.30769230769,
 'skoda': 110848.5639686684,
 'kia': 112530.30303030302,
 'toyota': 115944.35075885328,
 'land_rover': 118010.20408163265,
 'nissan': 118330.99579242637,
 'seat': 121131.30128956624,
 'jaguar': 124178.08219178082,
 'mazda': 124464.03385049365,
 'jeep': 127122.64150943396,
 'volkswagen': 128711.47607461477,
 'audi': 129157.38678544914,
 'mercedes_benz': 130788.36331334666,
 'bmw': 132575.9050214091,
 'volvo': 138067.9156908665}

In [51]:
brands_price_series = pd.Series(brands_price)
brands_mileage_series = pd.Series(brands_mileage)
brands_template = {'mean_price': brands_price_series, 'mean_mileage': brands_mileage_series}
brands_info = pd.DataFrame(data = brands_template)

In [52]:
brands_info

mean_price   mean_mileage
audi             9336.687454  129157.386785
bmw              8571.480148  132575.905021
chevrolet        6684.139098   99191.729323
dacia            5915.528455   84268.292683
hyundai          5365.254274  106442.307692
jaguar          11635.493151  124178.082192
jeep            11650.500000  127122.641509
kia              5982.330303  112530.303030
land_rover      19108.091837  118010.204082
mazda            4112.596615  124464.033850
mercedes_benz    8628.450366  130788.363313
mini            10613.459658   88105.134474
nissan           4743.402525  118330.995792
porsche         45643.937063   96853.146853
seat             4397.230950  121131.301290
skoda            6368.000000  110848.563969
sonstige_autos  23567.513043   89684.782609
toyota           5167.091062  115944.350759
volkswagen       5604.071269  128711.476075
volvo            4946.501171  138067.915691

There appears to not be any visible relation between the `mean_price` and `mean_mileage`. For this excercise, only visible analysis will be performed, given that its purpose is different from using medium to advanced statistical analysis. In a much complete exercise the appropriate approach would be to calculate a correlation coefficient.